lib/router/src/cache.rs

```diff
@@ -1,426 +1,367 @@
-
-        let t0 = clock.now();
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(333, MultiplyAndAssign::default());
-
-        clock.advance(Duration::from_secs(1));
-        assert_eq!(cache.access(&333).map(|n| n.last_access()), Some(t0));
-
-        // Cause the router to evict the `333` route.
-        clock.advance(Duration::from_secs(1));
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(444, MultiplyAndAssign::default());
-
-        clock.advance(Duration::from_secs(1));
-        let t1 = clock.now();
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(333, MultiplyAndAssign::default());
-
-        clock.advance(Duration::from_secs(1));
-        assert_eq!(cache.access(&333).map(|n| n.last_access()), Some(t1));
+            // Spawn a background purge
+            tokio::spawn(cache_purge);
+
+            {
+                let mut cache = cache.lock().unwrap();
+
+                cache.reserve().expect("reserve").store(1, 2);
+                cache.reserve().expect("reserve").store(2, 3);
+                assert_eq!(cache.vals.len(), 2);
+            }
+
+            // Sleep for an amount of time greater than expiry duration so
+            // that the background purge purges the entire cache.
+            tokio::spawn(
+                tokio_timer::sleep(Duration::from_millis(100))
+                    .map(move |_| {
+                        assert_eq!(cache.lock().unwrap().vals.len(), 0);
+                    })
+                    .map_err(|_| ()),
+            );
+
+            Ok(())
+        }))
     }
 
     #[test]
-    fn node_access_updated_on_drop() {
-        let mut clock = Clock::default();
-        let t0 = clock.now();
-        let mut node = Node::new(123, t0);
+    fn drop_access_resets_expiration() {
+        use tokio::runtime::current_thread::Runtime;
+
+        let mut runtime = Runtime::new().unwrap();
+
+        let (cache, cache_purge) = Cache::new(2, Duration::from_millis(10));
+
+        // Spawn a background purge on the current runtime
+        runtime.spawn(cache_purge);
 
-        clock.advance(Duration::from_secs(1));
         {
-            let access = node.access(&clock);
-            assert_eq!(access.last_access(), t0);
+            let mut cache_1 = cache.lock().unwrap();
+
+            cache_1.reserve().expect("reserve").store(1, 2);
+            assert!(cache_1.access(&1).is_some());
+            let _access = cache_1.access(&1);
+
+            // Sleep for an amount of time greater than the expiry duration so
+            // that the background purge would purge the cache iff values can
+            // be expired.
+            runtime
+                .block_on(tokio_timer::sleep(Duration::from_millis(100)))
+                .unwrap();
+
+            // Explicity drop both the access and cache handles. Dropping the
```

Comment:
Nit: I wouldn't say "Explicitly" here unless you actually call `drop(cache_1)` and so on...